### PR TITLE
perf: AsyncLocalStorageRunContextManager

### PR DESCRIPTION
**Pull request by trentm**
Implement a run-context manager using AsyncLocalStorage. Newer versions
of node (>=14.5, >=12.19) should benefit from significant better perf
compared to AsyncHooksRunContextManager, esp. for promise-heavy apps.

See the many async_hooks-related improvements in:
    https://nodejs.org/en/blog/release/v14.5.0/
    https://nodejs.org/en/blog/release/v12.19.0/

Refs: https://github.com/elastic/apm-agent-nodejs/issues/2745

### CODE REVIEW

**Comment by trentm** [comment]
@astorm Some review notes that might help. The initial work for this was to 3 files, but that become 27 files after config, docs, and test changes.

# Review notes

There are a few parts to this change that can mostly be reviewed separately:

- [ ] The implementation of the new `contextManager=asynclocalstorage|asynchooks|patch` config var that deprecates the `asyncHooks=true|false` config var. This part is all in:
    - lib/config.js
    - test/config.test.js
    - lib/instrumentation/index.js - The selection of the appropriate RunContextManager based on `contextManager`.
    - docs/configuration.asciidoc
- [ ] The addition of the `AsyncLocalStorageRunContextManager` class. The class hierarchy is:
    ```
    AbstractRunContextManager
        BasicRunContextManager (used for the non-async_hooks tracking)
            AsyncHooksRunContextManager
        AsyncLocalStorageRunContextManager
    ```
  The implementation of `.root()` was moved from `Basic...` up to `Abstract...`
  so that `AsyncLocalStorage...` could use it without using other unneeded
  parts of `Basic...`.
    - lib/instrumentation/run-context/AbstractRunContextManager.js
    - lib/instrumentation/run-context/BasicRunContextManager.js
  Then the implementation of `AsyncLocalStorage...` is very similar to OTel's
  `AsyncLocalStorageContextManager.ts`.
    - lib/instrumentation/run-context/AsyncLocalStorageContextManager.js
- [ ] The conversion of Jenkins CI to use `ELASTIC_APM_CONTEXT_MANAGER` rather than `ELASTIC_APM_ASYNC_HOOKS`.
    - .ci/...
- [ ] The conversion of the test suite to use `conf.contextManager` rather than `conf.asyncHooks`:
    - test/... (except "test/config.test.js")

File: lib/instrumentation/run-context/BasicRunContextManager.js [hunk 1]
```diff
@@ -7,31 +7,19 @@
 'use strict'
 
 const { AbstractRunContextManager } = require('./AbstractRunContextManager')
-const { RunContext } = require('./RunContext')
 
 // A basic manager for run context. It handles a stack of run contexts, but does
 // no automatic tracking (via async_hooks or otherwise). In combination with
-// "patch-async.js" it does an adequate job of context tracking for much of the
+// "patch-async.js" it does a limited job of context tracking for much of the
```
Comment (astorm) [inline lib/instrumentation/run-context/BasicRunContextManager.js:13]: 🔥 🔥 🔥 🔥

**Review by astorm — APPROVED** [review]
Code looks good and an `AsyncLocalStorage` option is a welcome change (especially with the potential perf. improvements for Kibana).  The class hierarchy makes sense and the configuration looks like it will correctly honor the `asyncHooks:false` for folks upgrading.

I don't have a lot of insight on the changes to the CI and what context managers are/aren't tested -- so if you have anything you're nervous about there you should probably get a second opinion but I don't have any objections to committing those changes.

File: docs/configuration.asciidoc [hunk 2]
```diff
@@ -226,20 +226,53 @@ If set to `true`, the client will poll the APM Server regularly for new agent co
...
+thread of control through asynchronous tasks. The preferred mechanism for this
+is https://nodejs.org/api/async_context.html#class-asynclocalstorage[`AsyncLocalStorage`],
+usable in Node.js versions >=14.5 and >=12.19. In older versions of Node, the APM
+agent will fallback to using
+https://nodejs.org/api/async_hooks.html[`async_hooks`], which can have a higher
+performance overhead, especially in Promise-heavy applications. Finally, the APM
+agent supports a limited tracking mechanism (named "patch") that monkey-patches
+much of the core Node.js API to attempt to follow asynchronous calls. The
+"patch" mechanism is limited in usage and coverage (for example, it cannot
+track `async/await` usage), is not recommended, and may be deprecated and
```
[the remark this answers is not in the document]
Comment (trentm) [reply docs/configuration.asciidoc]: @astorm I've done so in commit 8af97168. Does that commit look reasonable to you?

**Pull request merged**